pkg/operator/configobservation/encryption/key_controller_encryption.go

```diff
@@ -0,0 +1,246 @@
+package encryption
+
+import (
+	"crypto/rand"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const encWorkKey = "key"
+
+type EncryptionKeyController struct {
+	operatorClient operatorv1helpers.StaticPodOperatorClient
+
+	queue         workqueue.RateLimitingInterface
+	eventRecorder events.Recorder
+
+	preRunCachesSynced []cache.InformerSynced
+
+	validGRs map[schema.GroupResource]bool
+
+	componentName     string
+	componentSelector labels.Selector
+
+	secretLister corev1listers.SecretNamespaceLister
+	secretClient corev1client.SecretInterface
+}
+
+func NewEncryptionKeyController(
+	targetNamespace string,
+	operatorClient operatorv1helpers.StaticPodOperatorClient,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+	validGRs map[schema.GroupResource]bool,
+) *EncryptionKeyController {
+	c := &EncryptionKeyController{
+		operatorClient: operatorClient,
+		eventRecorder:  eventRecorder.WithComponentSuffix("encryption-key-controller"), // TODO unused
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "EncryptionKeyController"),
+
+		preRunCachesSynced: []cache.InformerSynced{
+			operatorClient.Informer().HasSynced,
+			kubeInformersForNamespaces.InformersFor(operatorclient.GlobalMachineSpecifiedConfigNamespace).Core().V1().Secrets().Informer().HasSynced,
```

Comment:
factor out `kubeInformersForNamespaces.InformersFor(operatorclient.GlobalMachineSpecifiedConfigNamespace)` into var and reuse below. Too easy to lose track of these, leading to ugly bugs.